### DISCUSSION TITLE
PutCompareExchengeValue after DeleteCompareExchangeValue

### DISCUSSION THREAD

**Discussion by ivanlegin** [ravendb/ravendb#19235] [Ideas]
Hello,

I've noticed a potential issue with compare exchange.

Specifically, when performing a Put operation, it will succeed if the data hasn’t been changed in the meantime. However, if the data has been deleted, the Put operation will succeed regardless of the index we pass as an argument.

I don’t think this is the best behavior because, for example, if we try to modify a document in a session with the retrieved change vector and the document has been deleted in the meantime, we will get a concurrency exception. Therefore, I believe the same logic should be applied to compareExchangeValue as well.

In other words, putCompareExchangeValue should not write if the index we passed as an argument does not match the one in the database, even if the record was deleted in the meantime. It should not accept any index, which is what it currently does.

In my opinion, it should only create a record if the passed index is 0, -1, or null when the document doesn’t exist.

Additionally, putCompareExchange should, if the document doesn’t exist and we passed an index other than 0, -1, or null, return one of these values to signal that the document doesn’t exist.

This could be problematic if we rely on the fetched value, make some changes with that data, or make a decision based on it, and then attempt to write it back. In the meantime, someone may have deleted the record, but putCompareExchangeValue will succeed because the document no longer exists, so it doesn't check the Index argument.

```
var first = await store.Operations.SendAsync(new PutCompareExchangeValueOperation<string>("key", 0, 0));
var second = await store.Operations.SendAsync(new DeleteCompareExchangeValueOperation<string>("key", first.Index));
// this value is no longer relevant because it was deleted in the meantime,
// but it will pass on the next put because when the document does not exist, 
// the index is not checked, and any index can be passed, 
// whereas in my opinion, it should only accept (0, -1, or null)
first.Value++;
var third = await store.Operations.SendAsync(new PutCompareExchangeValueOperation<string>("key", first.Value, first.Index));
if(!third.Successful && third.Index is 0 or -1 or null)
    PutCompareExchangeValueOperation<string>("key", 0, (0 or -1 or null)));
```

**Comment by karmeli87** [comment]
Hi,

We identified this as a problematic behavior as well https://issues.hibernatingrhinos.com/issue/RavenDB-22230, but we consider this as a breaking change and therefore this fix will be included only in the upcoming 6.1 version.
